### ISSUE TITLE
Ignore NULL values when reading maps

### ISSUE THREAD

**Issue by mernst-github** [pureconfig/pureconfig#1353]
When using layered configs it is useful to "null out" settings provided by the parent, e.g. with `reference.conf`:   `{a = a, b = b}` and `application.conf`:  `{b = null}`
with the intention of "making b go away" as if that mapping didn't exist at all.

It would be natural if pureconfig would ignore such entries instead of trying to make sense of them, that's the very intention of the null value in the config api (`hasPath("b")` is false). But instead it fails on them:

```
  "pureconfig" should {
    "ignore nulls" in {
      val config = ConfigFactory
        .parseString("{b = null}")
        .withFallback(ConfigFactory.parseString("{a = a, b = b}"))

      // b is not visible through the `Config` api
      config.entrySet() should be(Map("a" -> ConfigValueFactory.fromAnyRef("a")).asJava.entrySet())
      config.hasPath("b") should be(false)

      // but ConfigSource digs it out from the underlying `ConfigObject` and trips
      ConfigSource
        .fromConfig(config)
        .loadOrThrow[Map[String, String]] should be(Map("a" -> "a"))
    }
  }
```
=>
```
Cannot convert configuration to a scala.collection.immutable.Map. Failures are:
  at 'b':
    - (String: 1) Expected type STRING. Found NULL instead.
```

(pureconfig 0.17.1)

**Comment by ruippeixotog** [comment]
Hi @mernst-github! What you described is intended behavior on PureConfig.

While treating nulls as a way to "erase" existing values might be natural for you, many others might decide to treat a field explicitly defined as null differently from a missing one. PureConfig is generally more strict with types than Lightbend Config and much more aligned with Scala (where nulls are usually avoided).

Your config is more cleanly represented as a `Map[String, Option[String]]`, where PureConfig would read null values as `None`. This provides you maximum flexibility on how to treat keys defined in different forms and we recommend this approach.

That said, if you really need to read it as a `Map[String, String]`, you can easily override the existing behavior like this:

```scala
implicit def myCustomMapReader: ConfigReader[Map[String, String]] =
  ConfigReader[Map[String, Option[String]]].map(_.collect { case (k, Some(v)) => (k, v) })
```

**Comment by mernst-github** [comment]
Thanks for the response! I like that the custom reader you propose propagates also deep into subconfigs (wasn't sure whether that would work), so that is a feasible workaround also for nested config hierarchies. But I think it falls short for other cases where I'd find null = absence preferable, for example default values: `case class Config(a: String, b: String = "default")` cannot parse the above example, although there is an (imho obvious) interpretation.

The type systems of Config and Scala are obviously not a 1:1 match, so there are trade-offs to be made in the mapping, and pureconfig can choose to be helpful or strict^Wpicky, and it can distribute responsibility to the config author (_don't write nulls_) or the model author (_you have to use `Option` because the config author might use null overrides_) or find a middle ground.

In most cases, pureconfig chooses to be accomodating, e.g. by converting strings that look like numbers and vice versa and things "just work naturally". But in this case, different from the underlying Config, it chooses to be insistent that a `null` value in a config *is* a value and needs to be dealt with, instead of interpreting it as absent. I think that's the wrong choice.

A better approach imho is "if you want to be aware of nulls, use type `Option` and you'll see `None`. If you don't use `Option`, I'm not gonna bother you."

I think my best option to avoid this is to deep-copy the config upfront and drop all null mappings.
And the ironic part is that I don't even need to filter anything:
```
      val config = ...
      ConfigSource.fromConfig(
        ConfigFactory.parseMap {
          val m = new util.HashMap[String, ConfigValue]
          config.entrySet.forEach(e => m.put(e.getKey, e.getValue))
          m
        }
      )
```

**Comment by ruippeixotog** [comment]
Yeah, if you want null-as-undefined behavior on every config object regardless of the Scala type it reads into, I agree preprocessing the config is your best option. I think we could find ways to do it within PureConfig, involving pretty generic `ConfigReader`s and `.contramap` methods, but they'd be significantly more complex and wouldn't bring much value.

For context: PureConfig doesn't "choose to be accomodating" arbitrarily. The HOCON spec includes specific API recommendations about [automatic type conversions](https://github.com/lightbend/config/blob/4458ea947a7a2a668bb811a122455f1f05975172/HOCON.md#automatic-type-conversions) and PureConfig implements them. The spec doesn't say anything about having `null` as a value that should be ignored; on the contrary, the behavior of null in the "Substitutions" section explicitly says that it can neither be ignored nor treated as some kind of "delete" operation. For all the purposes in the spec, `null` is just a normal scalar value like `true` and `"abc"`.
